File: leetcode/python/977. 有序数组的平方.py
#!/bin/bash python
"""
977. 有序数组的平方

思路一：
暴力法，O(n + nlgn)

思路二：
二分查找找到负数和非负数的区分点
然后从中间向两头归并
也就是从最小值往最大值归并

思路三：
直接从两头向中间归并（厉害了，要去除思维定式）
也就是从最大值往最小值归并
"""

class Solution(object):
    def sortedSquares(self, a):
        """
        暴力法
        """
        ret = [x ** 2 for x in a]
        return sorted(ret)

    def sortedSquares2(self, a):
        """
        二分+归并
        """
        # 输出二分查找分界点l，l及其右侧都是非负数
        # 特殊情况，l遍历到右边界需要处理
        l, h = 0, len(a) - 1
        while l < h:
            m = l + (h - l) // 2
            if a[m] < 0:
                l = m + 1
            elif a[m] >= 0:
                h = m
        # 归并合并a[0:m-1], a[m:]，注意m有可能等于h
        m = l + 1 if a[l] < 0 else l
        p, q = m - 1, m
        # 归并合并
        ret = [0] * len(a)
        for i in range(len(a)):
            if p >= 0 and (q > len(a) - 1 or a[p] ** 2 < a[q] ** 2):
                ret[i] = a[p] ** 2
                p -= 1
            else:
                ret[i] = a[q] ** 2
                q += 1
        return ret

    def sortedSquare3(self, a):
        """
        直接归并
        """
        ret = a[:]
        p, q = 0, len(a) - 1
        for i in range(len(a) - 1, -1, -1):
            if p <= q and a[p] ** 2 >= a[q] ** 2:
                ret[i] = a[p] ** 2
                p += 1
            else:
                ret[i] = a[q] ** 2
                q -= 1
        return ret

    def test(self, lst):
        """
        测试函数
        """
        func_vec = [self.sortedSquares,
                    self.sortedSquares2,
                    self.sortedSquare3]
        for a in lst:
            print("=====")
            print("input data: {}".format(a))
            for fun in func_vec:
                print(fun(a))

if __name__ == "__main__":
    s = Solution()
    lst = [[-4,-1,0,3,10],
           [-7,-3,2,3,11],
           [-7,-3,-2],
           [2,3,11]]
    s.test(lst)

File: leetcode/cpp/977. 有序数组的平方.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>

class Solution {
public:
    // 暴力法
    std::vector<int> sortedSquares(std::vector<int>& a) {
        std::vector<int> ret;
        for (auto & i : a) {
            ret.push_back(i * i);
        }
        sort(ret.begin(), ret.end());
        return ret;
    }

    // 二分返回负数和非负数的切分点
    int divideIndex(std::vector<int>& a) {
        int l = 0;
        int h = a.size() - 1;
        while (l < h) {
            int m = l + (h - l) / 2;
            if (a[m] < 0) {
                l = m + 1;
            } else if (a[m] >= 0) {
                h = m;
            }
        }
        return (a[l] < 0 ? l + 1 : l);
    }

    // 二分 + 归并
    std::vector<int> sortedSquares2(std::vector<int>& a) {
        int m = divideIndex(a);
        int p = m - 1;
        int q = m;
        std::vector<int> ret(a.size(), 0);
        for (int i = 0; i < ret.size(); i++) {
            if (p >= 0 && (q > a.size() - 1 || a[p]*a[p] < a[q]*a[q])) {
                ret[i] = a[p] * a[p];
                p -= 1;
            } else {
                ret[i] = a[q] * a[q];
                q += 1;
            }
        }
        return ret;
    }

    // 直接从两端归并
    std::vector<int> sortedSquares3(std::vector<int>& a) {
        int p = 0;
        int q = a.size() - 1;
        std::vector<int> ret(a.size(), 0);
        for (int i = ret.size() - 1; i >= 0; i--) {
            if (p <= q && a[p]*a[p] >= a[q]*a[q]) {
                ret[i] = a[p] * a[p];
                p += 1;
            } else {
                ret[i] = a[q] * a[q];
                q -= 1;
            }
        }
        return ret;
    }

    // 打印一维数组
    void printf_1d_vec(std::vector<int>& a) {
        for (auto & i : a) {
            printf("%d, ", i);
        }
        printf("\n");
    }

    // 测试函数
    void test(std::vector<std::vector<int>>& lst) {
        for (auto & a : lst) {
            printf("=====\n");
            printf("input data: ");
            printf_1d_vec(a);
            for (auto & func : this->func_vec_) {
                std::vector<int> result = (this->*func)(a);
                printf_1d_vec(result);
            }
        }
    }

private:
    typedef std::vector<int> (Solution::*func_ptr)(std::vector<int>&);
    std::vector<func_ptr> func_vec_ = {&Solution::sortedSquares,
                                       &Solution::sortedSquares2,
                                       &Solution::sortedSquares3};
};

int main(int argc, char* argv[])
{
    std::vector<std::vector<int>> lst = {
        {-4,-1,0,3,10},
        {-7,-3,2,3,11},
        {-7,-3,-2},
        {2,3,11}
    };
    Solution s;
    s.test(lst);
    return 0;
}